File: cmd/examples/hello-ruby/hello.rb
#!/usr/bin/env ruby

require 'webrick'
require 'optparse'

# Parse command line arguments
options = { listen: 'localhost:8080' }
OptionParser.new do |opts|
  opts.banner = "Usage: hello.rb [options]"

  opts.on("--listen ADDR", "IP:port to listen on (default: localhost:8080)") do |addr|
    options[:listen] = addr
  end
end.parse!

# Parse listen address
listen_addr = options[:listen]
host = ''
port = 8080

if listen_addr.start_with?(':')
  port = listen_addr[1..-1].to_i
elsif listen_addr.include?(':')
  parts = listen_addr.rpartition(':')
  host = parts[0]
  port = parts[2].to_i
else
  port = listen_addr.to_i
end

# Get client IP from request
def get_client_ip(req)
  # Try to get IP from X-Forwarded-For header (for proxies)
  if req.header['x-forwarded-for']&.any?
    forwarded = req.header['x-forwarded-for'].first
    ips = forwarded.split(',')
    return ips[0].strip if ips.any?
  end

  # Try to get IP from X-Real-IP header
  if req.header['x-real-ip']&.any?
    return req.header['x-real-ip'].first
  end

  # Fall back to remote address
  req.peeraddr[3]
end

# Create HTTP server
server = WEBrick::HTTPServer.new(
  BindAddress: host.empty? ? nil : host,
  Port: port,
  Logger: WEBrick::Log.new($stdout, WEBrick::Log::INFO),
  AccessLog: []
)

# Define request handler
server.mount_proc '/' do |req, res|
  # Check if required headers exist
  login_name = req.header['tailscale-user-login']&.first
  display_name = req.header['tailscale-user-name']&.first
  profile_pic_url = req.header['tailscale-user-profile-pic']&.first

  res['Cache-Control'] = 'no-cache'

  if login_name && display_name && profile_pic_url
    # Serve HTML page with user information
    res['Content-Type'] = 'text/html'
    res.body = <<~HTML
      <!DOCTYPE html>
      <html>
      <head>
          <title>Hello from Ruby!</title>
      </head>
      <body>
          <h1>(Ruby) Tailscale User Information</h1>
          <p><strong>Login Name:</strong> #{login_name}</p>
          <p><strong>Name:</strong> #{display_name}</p>
          <p><strong>Profile Picture:</strong></p>
          <img src="#{profile_pic_url}" alt="Profile Picture" style="max-width: 200px;">
      </body>
      </html>
    HTML
  else
    # Print anonymous message with IP
    ip = get_client_ip(req)
    res['Content-Type'] = 'text/plain'
    res.body = "Hello anonymous from #{ip}\n"
  end
end

# Handle graceful shutdown
trap('INT') do
  puts "\nShutting down server..."
  server.shutdown
end

display_host = host.empty? ? '0.0.0.0' : host
puts "Starting server on #{display_host}:#{port}"

server.start

File: cmd/examples/hello-perl/hello.pl
#!/usr/bin/env perl

use strict;
use warnings;
use HTTP::Daemon;
use HTTP::Status;
use Getopt::Long;

# Parse command line arguments
my $listen_addr = 'localhost:8080';
GetOptions('listen=s' => \$listen_addr) or die "Usage: $0 [--listen address:port]\n";

# Parse listen address
my ($host, $port) = ('', 8080);
if ($listen_addr =~ /^:(\d+)$/) {
    $port = $1;
} elsif ($listen_addr =~ /^(.+):(\d+)$/) {
    $host = $1;
    $port = $2;
} elsif ($listen_addr =~ /^(\d+)$/) {
    $port = $1;
}

# Default to localhost if not specified
$host = 'localhost' if $host eq '';

# Get client IP from headers
sub get_client_ip {
    my ($req) = @_;

    # Try X-Forwarded-For first
    my $forwarded = $req->header('X-Forwarded-For');
    if ($forwarded) {
        my @ips = split /,/, $forwarded;
        if (@ips) {
            $ips[0] =~ s/^\s+|\s+$//g;
            return $ips[0];
        }
    }

    # Try X-Real-IP
    my $real_ip = $req->header('X-Real-IP');
    return $real_ip if $real_ip;

    # Fall back to peer address
    return 'unknown';
}

# Create HTTP daemon
my $daemon = HTTP::Daemon->new(
    LocalAddr => $host,
    LocalPort => $port,
    ReuseAddr => 1,
) or die "Cannot create HTTP daemon: $!";

print STDERR "Starting server on $host:$port\n";

# Handle Ctrl+C gracefully
$SIG{INT} = sub {
    print STDERR "\nShutting down server...\n";
    exit 0;
};

# Main server loop
while (my $client = $daemon->accept) {
    while (my $req = $client->get_request) {
        # Check for Tailscale headers
        my $login_name = $req->header('Tailscale-User-Login');
        my $display_name = $req->header('Tailscale-User-Name');
        my $profile_pic_url = $req->header('Tailscale-User-Profile-Pic');

        # Create response
        my $res = HTTP::Response->new(HTTP::Status::RC_OK);
        $res->header('Cache-Control' => 'no-cache');

        if ($login_name && $display_name && $profile_pic_url) {
            # Serve HTML page with user information
            $res->header('Content-Type' => 'text/html');
            my $body = <<"HTML";
<!DOCTYPE html>
<html>
<head>
    <title>Hello from Perl!</title>
</head>
<body>
    <h1>(Perl) Tailscale User Information</h1>
    <p><strong>Login Name:</strong> $login_name</p>
    <p><strong>Name:</strong> $display_name</p>
    <p><strong>Profile Picture:</strong></p>
    <img src="$profile_pic_url" alt="Profile Picture" style="max-width: 200px;">
</body>
</html>
HTML
            $res->content($body);
        } else {
            # Print anonymous message with IP
            my $client_ip = get_client_ip($req);
            $res->header('Content-Type' => 'text/plain');
            $res->content("Hello anonymous from $client_ip\n");
        }

        # Send response
        $client->send_response($res);
    }

    $client->close;
    undef $client;
}
